src/daft-scan/src/scan_task_iters/split_parquet_files_by_rowgroup.rs

```diff
@@ -0,0 +1,285 @@
+                Ok(scan_task_to_split) => {
+                    EitherOnceOrMany::Many(scan_task_to_split.into_iter(config).map(Ok))
+                }
+            }),
+    ) as BoxScanTaskIter
+}
+
+/// A ScanTask that can be split by Parquet rowgroups
+pub(crate) enum SplittableScanTaskRef {
+    NoSplit(ScanTaskRef),
+    Split(ScanTaskRef, FileMetaData),
+}
+
+/// Iterator for [`SplittableScanTaskRef`] which produces [`ScanTaskRef`]
+enum SplittableScanTaskRefIterator<'a> {
+    NoSplitIterOnce(ScanTaskRef, bool),
+    SplitIterMany(ParquetFileSplitter<'a>),
+}
+
+/// Iterator that performs a split on the provided [`ScanTaskRef`]
+struct ParquetFileSplitter<'a> {
+    // Iterator state that tracks iteration progress
+    current_idx: usize,
+    accumulated_rowgroup_ids: Vec<usize>,
+    estimated_accumulated_materialized_size: f64,
+
+    // Cached items
+    scan_task: ScanTaskRef,
+    file_metadata: FileMetaData,
+    config: &'a DaftExecutionConfig,
+
+    // Call once and cache this, since it can be expensive when called in a loop
+    scan_task_in_memory_size_estimate: Option<usize>,
+}
+
+impl SplittableScanTaskRef {
+    fn into_iter(self, cfg: &DaftExecutionConfig) -> SplittableScanTaskRefIterator {
+        match self {
+            Self::NoSplit(st) => SplittableScanTaskRefIterator::NoSplitIterOnce(st, false),
+            Self::Split(st, fm) => {
+                let scan_task_in_memory_size_estimate = st.estimate_in_memory_size_bytes(Some(cfg));
+                SplittableScanTaskRefIterator::SplitIterMany(ParquetFileSplitter {
+                    current_idx: 0,
+                    accumulated_rowgroup_ids: Vec::new(),
+                    estimated_accumulated_materialized_size: 0.,
+                    scan_task: st,
+                    file_metadata: fm,
+                    config: cfg,
+                    scan_task_in_memory_size_estimate,
+                })
+            }
+        }
+    }
+}
+
+impl<'a> Iterator for SplittableScanTaskRefIterator<'a> {
+    type Item = ScanTaskRef;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self {
+            Self::NoSplitIterOnce(st, has_returned) => {
+                if !*has_returned {
+                    *has_returned = true;
+                    Some(st.clone())
+                } else {
+                    None
+                }
+            }
+            Self::SplitIterMany(splitter) => splitter.next(),
+        }
+    }
+}
+
+impl<'a> Iterator for ParquetFileSplitter<'a> {
+    type Item = ScanTaskRef;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while !self.exhausted() {
+            let emitted_result = self.advance_state();
+            if let Some(emitted_result) = emitted_result {
+                return Some(emitted_result);
+            }
+        }
+        None
+    }
+}
+
+impl<'a> ParquetFileSplitter<'a> {
+    /// Returns whether or not this Splitter is done producing items
+    pub fn exhausted(&self) -> bool {
```

Comment:
why are these pub? 

shouldn't these be fine to have the visibility to the module (default)?

Reply:
These will be limited by the visibility of the struct `ParquetFileSplitter` itself. It was previously private, so it wouldn't have had visibility beyond the module.

I had to refactor `ParquetFileSplitter` to be `pub(crate)` though, because it is now exposed through the implementation of `IntoIterator` as per https://github.com/Eventual-Inc/Daft/pull/3454#discussion_r1884381990, so note that these will not have `pub(crate)` visibility. Updated.